core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -78,7 +78,8 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
     ZIO.async[Any, Nothing, Exit[E, A]](
       cb =>
         tell(FiberMessage.Stateful { (fiber, _) =>
-          if (fiber._exitValue ne null) cb(Exit.Success(fiber.exitValue()(Unsafe.unsafe).asInstanceOf[Exit[E, A]]))
+          if (fiber._exitValue ne null)
+            fiber.removeObserver(exit => cb(Exit.Success(exit.asInstanceOf[Exit[E, A]])))(Unsafe.unsafe)
```

Comment:
The callback is never going to be called if the fiber is already completed. Also, removing the callback is not going to do anything because we just created it so no one else has a reference to it so there is nothing to remove.